### DISCUSSION TITLE
Pandora's Tower scaled EFB copy Bloom, DOF and double vision issue

### DISCUSSION THREAD

**Discussion by ZexRon32** [TryTwo/dolphin#9] [Q&A]
Hello, I made a Reddit post regarding the bloom issue with Pandora's Tower : [https://www.reddit.com/r/DolphinEmulator/comments/1fj2xyk/pandoras_tower_scaled_efb_copy_bloom_dof_and/](https://www.reddit.com/r/DolphinEmulator/comments/1fj2xyk/pandoras_tower_scaled_efb_copy_bloom_dof_and/)

BloomControl doesn't work properly for Pandora's Tower. It seems to be unable target the bloom unless you select the full range of EFB copies. This includes lighting, color correction, bloom, depth of field, reflections, and an unknown rectangle on the top left of the screen. 

![Scaled EFB copy Exclusions 50% bightness](https://github.com/user-attachments/assets/1a46cec8-789e-4ad1-bcbb-c3ef7e032d10)

If you please visit my post to have more detail and eventually find a solution. Thank you.

**Comment by TryTwo** [comment]
Can you send a 1-frame fifo recording?  Tools -> Fifo Player. 

You could also try this Graphics Mod branch that uses a different method to try and fix it: https://github.com/dolphin-emu/dolphin/pull/12280

At the bottom, "show all checks" -> the one that fits your system (pr-win-x64 for windows), -> the "upload" link at the bottom of that page.

**Reply by ZexRon32** [reply]
I put the fifo recording on the goole drive : [Scaled EFB copy Exclusions 50% bightness 1-frame.dff](https://drive.google.com/file/d/1HPcecpwbvVpkaI5uUaSCUJcQ604VPoie/view?usp=sharing)
As for the graphics mod editor, I don't think I'm competent enough to find a way to fix it. I'll update my Reddit post if I find something that can help. Let me know if you want me to update you.
Thank

**Reply by TryTwo** [reply]
The editor already has the method built in.  You should just need to play with the normal graphics mod options for the game while using that build.

**Reply by ZexRon32** [reply]
The graphics mods don't seem to work properly for Pandora's Tower
I found the ObjectID that is drawing the bloom using the graphics mod editor : 1546742346967714003
I'm able to skip drawing this object to disable the bloom. Still, it would be great for it to render with blur.
![image](https://github.com/user-attachments/assets/1715a6ad-5f87-4593-af7f-0db3af06c9ae)

**Reply by TryTwo** [reply]
Ha, I just did the same thing. It almost looks like it's doing a drop-shadow effect.   I noticed you can apply a transform, rather than skip, to what you found to get it to center better.  -2.0 in the first box, 0 for the rest worked for me. Hard to tell with just 1-frame if it's a solution.

My bloom mod build just can't affect this game. I even tried to force it, but nope.  The new graphics mod has a feature to blur stuff too, but it's not in the UI.  I think you can blur it by adding the number you found to: open the graphicsmod custom dolphin version you downloaded -> sys -> load -> graphics mods -> pandora's tower -> metadata. Update the targets section like so:

```  "targets": [
    {
      "id": "10512703869395082884",
      "name": "",
      "tags": [
        "Bloom"
      ],
      "type": "int"
    },
    {
      "id": "1546742346967714003",
      "name": "",
      "tags": [
        "Bloom"
      ],
      "type": "int"
    }
```

Then I think you go to the game properties and enable native res bloom (which should actually blur it and not downscale).

**Reply by iwubcode** [reply]
TryTwo is correct.  That is certainly one way to do it.  There is also a way to do it in the editor that should work (create a new project and then tag the draw as "Bloom", then save the project), this will leave the system copy alone.

My only concern is whether this will work for Pandora's Tower.  The screenshot indicates that the game combines a color copy with the actual bloom metadata.  If I'm correct, the current implementation will blur _everything_ which isn't what we want.  I have seen other games that opted that way (ex: Rune Factory Frontier), which is why I have not released this feature yet.

I am working on a minor rewrite of the material system to make it a bit easier to tweak the Dolphin output but we might need some additional features.  The hope is to support all bloom games and have users just tag the bloom copies in the editor.

**Reply by ZexRon32** [reply]
To answer @iwubcode . Tagging the object as "bloom" as no visual effect.
I was unable to find the "Pandoras Tower Definitions" graphic mod in the list even though I found it in the files. All other graphics mods made by the dolphin Team as no visual effect at first, then crashes the custom dolphin build when you disable them. So "Native Resolution Bloom" or "Native Resolution DOF" can't be enabled. I don't know where the crash logs are located. I found a "dolphin.log", but it's empty.
I made my own graphic mod using @TryTwo advice : [Bloom Alignement](https://drive.google.com/drive/folders/18ENkZ4KmuiOAgJ6SdFAvT_lsczTdu4YM?usp=sharing)
It tags object "1546742346967714003" (Game Bloom) and "18112275771881123146" (Main menu Bloom) as bloom and translate them by -2. I left object "10512703869395082884" tagged as bloom, but I don't really know what it is, because it just makes the image brighter when skip.

The result :
![SX3P01_2024-09-18_15-03-04](https://github.com/user-attachments/assets/a7e0faf8-20a1-41fa-a706-a1822d8e1b79)

**Reply by iwubcode** [reply]
>I was unable to find the "Pandoras Tower Definitions" graphic mod

The definitions are combined with the actual mod (remove or native resolution bloom) to achieve the required effect.  The definitions are just which draw calls are tagged as bloom.  That listing won't show up in the menu.

> Tagging the object as "bloom" as no visual effect.

If you just tagged in the editor and expected things to work, it won't.  This is because I don't combine the tag with the system level details currently.  You have to save your project and then disable the editor and use the mod directly.  It's a pain and will be fixed before the real release of the editor.

If you weren't testing in editor, then maybe you are running into a bug.  I only tested a handful of games with the new system.  I'll be sure to do an exhaustive test once things are further along.

> All other graphics mods made by the dolphin Team as no visual effect at first, then crashes the custom dolphin build when you disable them

Not sure which mod you are talking about but I haven't converted over all of the bloom mods to the new graphics mods 2.0 format.  And definitely not the DOF mods.  Many games will be broken.  They will be fixed once the functionality I mentioned previously is implemented.

>It tags object "1546742346967714003" (Game Bloom) and "18112275771881123146" (Main menu Bloom) as bloom and translate them by -2.

I'm not sure what the translation is intended to do but I'd assume you don't need the tags if you are doing that.  The tags are what tell Dolphin to run the blur operation.  This will slow down performance so it's up to you to decide if you need it.

>The result :

Not sure if this is what you wanted or not but appreciate you playing around with the editor :)

**Reply by TryTwo** [reply]
I think setting dolphin to portable with the portable.txt file will get things to show up in the menus.  Otherwise I don't think it will use the sys folder inside the game directory?  I haven't played around with it for awhile though.

I'm also curious if those results are good or if there are still issues. The SS looks better than before.

**Reply by ZexRon32** [reply]
> Not sure which mod you are talking about but I haven't converted over all of the bloom mods to the new graphics mods 2.0 format.

I was talking about "Bloom Removal", "DOF removal" and "HUD removal". When enable, there's no visual change.
In the case of "Native Resolution Bloom", I get an error :
<img width="408" alt="Dolphin_TOBOmYRnas" src="https://github.com/user-attachments/assets/305e4fa2-9d16-4be3-8489-59a77f40f9c1">
Here's the log file : [bad_ps_Vulkan_0.txt](https://github.com/user-attachments/files/17062475/bad_ps_Vulkan_0.txt)
It's the same with "Native Resolution DOF" but it's to be expected because you said it's broken in most games.

>If you just tagged in the editor and expected things to work, it won't. This is because I don't combine the tag with the system level details currently. You have to save your project and then disable the editor and use the mod directly.

By disabling the editor, do you mean closing the project? If not, I don't know how to disable it. For me, there's always the editor's bar on top of the game windows, even if there's no project opened. I've exported a scene metadata as json where the bloom is tagged as such. At first it didn't appear in the graphics mods list. I had to set an action (translate an object by -0 which does nothing) to make it appear, so I can enable it. I didn't see a visual change.

> I'm not sure what the translation is intended to do but I'd assume you don't need the tags if you are doing that.

It aligns the bloom object with the game image to prevent the double vision effect. It makes it look like the bloom is disabled.

>Not sure if this is what you wanted

> I'm also curious if those results are good or if there are still issues.

It's so much better that it was before. Its best result for now. Still, I don't think it's the intended one. What would be the best is the result I was able to get with Zelda Twilight Princess using Bloom Control:

Good Bloom :
![RZDP01_2024-09-19_18-54-54](https://github.com/user-attachments/assets/49f7387a-37c7-4fe4-a692-050d4beb3430)
Bad Bloom :
![RZDP01_2024-09-19_16-17-14](https://github.com/user-attachments/assets/f24098f4-ca9d-43bb-8a81-864d6a686b3a)

I will continue to play with the graphic mod editor because it seems to be a very powerful tool, even if it doesn't seem to be able to have the result I wanted. I'm Looking forward to your future release.

**Reply by iwubcode** [reply]
>In the case of "Native Resolution Bloom", I get an error

Apologies for this.  There have been a lot of changes as of late and looks like I forgot to update the shader after I made a recent change.  I didn't realize I broke that.  I'll be sure to test this before my next push.

> By disabling the editor, do you mean closing the project?

Yep, that is correct.  Sounds like you did nothing wrong.

>I had to set an action (translate an object by -0 which does nothing) to make it appear, so I can enable it

Hmm, that shouldn't be necessary.  I will look into it when I get some time.  Thanks!

> It aligns the bloom object with the game image to prevent the double vision effect. It makes it look like the bloom is disabled.

Ah ok, that makes sense.  I guess slightly different result than outright removing bloom.  Thanks for explaining that.

> It's so much better that it was before. Its best result for now. Still, I don't think it's the intended one. What would be the best is the result I was able to get with Zelda Twilight Princess using Bloom Control

Glad to hear it is better but yes, that TP output is the desired result.  Before I broke the shader, I had something similar in TP, though the bloom was not that orange-y color, that led me to my current work.

>I will continue to play with the graphic mod editor because it seems to be a very powerful tool, even if it doesn't seem to be able to have the result I wanted. I'm Looking forward to your future release.

Thank you very much for all the testing and the kind words!  I too will be excited when this releases.  Still a bit more work to do :)
